sql: validate function leakproof volatility at builder level.

chengxiong-ruan opened the pull request:
Fixes: #97427

Previously, in declarative schema, we count on descriptor self validation to error out on invalid function volatility with leakproof flag as true. This is bad because descriptor validation panics at execution time which error out with a whole big print of the plan. This commit add such validation in the builder level to avoid such user unreadable error.

Release note: None

chengxiong-ruan committed:
sql: validate function leakproof volatility at builder level.

Fixes: #97427

Previously, in declarative schema, we count on descriptor self validation
to error out on invalid function volatility with leakproof flag as true.
This is bad because descriptor validation panics at execution time which
error out with a whole big print of the plan. This commit add such validation
in the builder level to avoid such user unreadable error.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/97778)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
> Reviewed 1 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97778)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_function.go` line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97778#-NPLDJ9hEr4HBmcpKLEs:-NPLDJ9hEr4HBmcpKLEt:b6i0xat) ([raw file](https://github.com/cockroachdb/cockroach/blob/61392fab5fc13cc30f0b2986139f31148ff5bca0/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_function.go#L200)):*
> ```Go
> 			pgerror.Newf(
> 				pgcode.InvalidFunctionDefinition,
> 				"cannot set leakproof on function with non-immutable volatility: %s",
> ```

make an error in `sqlerrors`, also, please add the same logical validation during the execution of the legacy schema changer, if you can, and make the error in validation not have a pgcode. Pgcodes in validation errors are an anti-pattern. 



_Code quote:_
```Go
cannot set leakproof on function
```
___
*[`pkg/sql/logictest/testdata/logic_test/udf` line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97778#-NPLDCIH9tf1H3Wk0FwE:-NPLDCIH9tf1H3Wk0FwF:b-rj7qbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/61392fab5fc13cc30f0b2986139f31148ff5bca0/pkg/sql/logictest/testdata/logic_test/udf#L25)):*
> ```
> $$ LANGUAGE made_up_language
> 
> statement error pq: cannot set leakproof on function with non-immutable volatility: STABLE
> ```

can you assert the pg code?


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97778)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_function.go` line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97778#-NPLDJ9hEr4HBmcpKLEs:-NPLLz0jDvMtt90XZ0IP:bpfrwdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/61392fab5fc13cc30f0b2986139f31148ff5bca0/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_function.go#L200)):*
In legacy schema changer just conveniently do the validation on the descriptor (not in descriptor validation). But yeah, I'll change it so the logical would look similar.

> Pgcodes in validation errors are an anti-pattern.

TIL.. Thanks for pointing that out.
___
*[`pkg/sql/logictest/testdata/logic_test/udf` line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97778#-NPLDCIH9tf1H3Wk0FwE:-NPLMJd30Nu9JZKmXSfZ:bjq966u) ([raw file](https://github.com/cockroachdb/cockroach/blob/61392fab5fc13cc30f0b2986139f31148ff5bca0/pkg/sql/logictest/testdata/logic_test/udf#L25)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can you assert the pg code?
</blockquote></details>

Won't need to assert pg code since I'll remove pgcode from the error.


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
> I was saying that this should have a pgcode, and it should not come from validation code.

Sorry I misunderstood. I think I addressed your comments.

ajwerner commented on pkg/sql/catalog/catpb/function.go:
```diff
@@ -0,0 +1,14 @@
+
+package catpb
+
+const DefaultFunctionVolatility = Function_VOLATILE
```
nit: comments

chengxiong-ruan replied:
done.

ajwerner commented on pkg/sql/catalog/funcdesc/func_desc.go:
```diff
@@ -836,14 +836,36 @@ func IsOIDUserDefinedFunc(oid oid.Oid) bool {
+		case tree.FunctionVolatility:
+			v, err := VolatilityToProto(t)
+			if err != nil {
+				panic(err)
```
Why panic?

chengxiong-ruan replied:
huh, forgot to change to it to return error after moving it out of declarative schema changer.
Fixed.

ajwerner commented on pkg/sql/catalog/funcdesc/func_desc.go:
```diff
@@ -836,14 +836,36 @@ func IsOIDUserDefinedFunc(oid oid.Oid) bool {
+	}
+
+	if err := mustMatchLeakProofVolatility(curLeakProof, curVolatility); err != nil {
+		return pgerror.Wrap(err, pgcode.InvalidFunctionDefinition, "invalid volatility")
```
I think my preference would be to have this return an unwrapped error and then wrap the pgcode above.

chengxiong-ruan replied:
Make sense. It is weird returning pg errors from this package.

ajwerner commented on pkg/sql/catalog/funcdesc/func_desc.go:
```diff
@@ -836,14 +836,36 @@ func IsOIDUserDefinedFunc(oid oid.Oid) bool {
+		return errors.Newf(
 			"cannot set leakproof on function with non-immutable volatility: %s",
-			fn.GetVolatility().String(),
+			vol.String(),
```
Why are we calling the string method?

chengxiong-ruan replied:
no need indeed, removed.

ajwerner reviewed (commented):
I've got some more feedback on how I think this could be a bit cleaner. WDYT?

> Reviewed 1 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97778#-:-NPP2OHB0tMHt9UK-MvO:bmw36e1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan)

___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 849 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97778#-NPL3h8e66GgAouVVViV-r2-849:-NPOzQ2MCAqM52Cpsq7X:b-6j6ikp) ([raw file](https://github.com/cockroachdb/cockroach/blob/089b7b179dcc5bc440b18ebc00d6651ba76ae718/pkg/sql/catalog/funcdesc/func_desc.go#L849)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

huh, forgot to change to it to return error after moving it out of declarative schema changer.
Fixed.
</blockquote></details>

nit: should we wrap this with an assertion failure?
___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 839 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/97778#-NPP-2Kx5SNyHp8fxQuc:-NPP-2Ky2Y02C9ti689W:bjpv631) ([raw file](https://github.com/cockroachdb/cockroach/blob/4497047ec68b04c9dd1141a6dbc66684ed8551c2/pkg/sql/catalog/funcdesc/func_desc.go#L839)):*
> ```Go
> // ValidateLeakProofVolatility returns an error when a function is defined as
> // leakproof but not immutable. See more details in comments for volatility.V.
> func ValidateLeakProofVolatility(
> ```

the arguments to this function are sort of confusing. Can you say that the arguments are the assumed defaults if the options do not specify? `cur` is not a super obvious prefix to me given this is a top-level function.

To be honest, it feels like what should really be going on here is a function like:

```
package catalog // I'd also be cool with a new package like `funcinfo` in catalog

// VolatilityProperties are properties about the volatility of user-defined functions.
type VolatilityProperties struct {
    LeakProof bool
    Volatility catpb.Function_Volatility
}

// ...
func GetVolatilityPropertiesFromFunctionOptions(
   options tree.FunctionOptions, 
) (VolatilityProperties, error) {
   v := VolatilityProperties{
    LeakProof: catpb.DefaultFunctionLeakProof,
    Volatilit: catpb.DefaultFunctionVolatility,
   }
   
    // ...
}

// ...
func (v VolatilityProperties) Valid() bool {
    return !leakProof || volatility == catpb.Function_IMMUTABLE
}

```

```
package sqlerrors

func NewVolatilityLeakProofMismatchError(...) error
```

Then, in sql, this can be like:

```
volatilityProps, err := catalog.GetVolatilityPropertiesFromFunctionOptions(options)
if err != nil {
    return err
}
if !volatilityProps.Valid() {
   return ..
}
```
___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 865 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/97778#-NPP0_Yt45mP24xA6WYi:-NPP0_Yt45mP24xA6WYj:b6b14c8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4497047ec68b04c9dd1141a6dbc66684ed8551c2/pkg/sql/catalog/funcdesc/func_desc.go#L865)):*
> ```Go
> 	if leakProof && vol != catpb.Function_IMMUTABLE {
> 		return errors.Newf(
> 			"cannot set leakproof on function with non-immutable volatility: %s",
> ```

I think this is a bad message for the validation error -- nothing is being set.


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97778)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @msirek)

___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 849 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97778#-NPL3h8e66GgAouVVViV-r2-849:-NPTSqOTB61uvqSfEpLe:b-qu6n94) ([raw file](https://github.com/cockroachdb/cockroach/blob/089b7b179dcc5bc440b18ebc00d6651ba76ae718/pkg/sql/catalog/funcdesc/func_desc.go#L849)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: should we wrap this with an assertion failure?
</blockquote></details>

Good point. Done.
___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 839 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/97778#-NPP-2Kx5SNyHp8fxQuc:-NPTScS3ESrhbEy3FhfD:b-58xa2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/4497047ec68b04c9dd1141a6dbc66684ed8551c2/pkg/sql/catalog/funcdesc/func_desc.go#L839)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

the arguments to this function are sort of confusing. Can you say that the arguments are the assumed defaults if the options do not specify? `cur` is not a super obvious prefix to me given this is a top-level function.

To be honest, it feels like what should really be going on here is a function like:

```
package catalog // I'd also be cool with a new package like `funcinfo` in catalog

// VolatilityProperties are properties about the volatility of user-defined functions.
type VolatilityProperties struct {
    LeakProof bool
    Volatility catpb.Function_Volatility
}

// ...
func GetVolatilityPropertiesFromFunctionOptions(
   options tree.FunctionOptions, 
) (VolatilityProperties, error) {
   v := VolatilityProperties{
    LeakProof: catpb.DefaultFunctionLeakProof,
    Volatilit: catpb.DefaultFunctionVolatility,
   }
   
    // ...
}

// ...
func (v VolatilityProperties) Valid() bool {
    return !leakProof || volatility == catpb.Function_IMMUTABLE
}

```

```
package sqlerrors

func NewVolatilityLeakProofMismatchError(...) error
```

Then, in sql, this can be like:

```
volatilityProps, err := catalog.GetVolatilityPropertiesFromFunctionOptions(options)
if err != nil {
    return err
}
if !volatilityProps.Valid() {
   return ..
}
```

</blockquote></details>

Thanks for the suggestion. I changed it as mostly like what you proposed.
___
*[`pkg/sql/catalog/funcdesc/func_desc.go` line 865 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/97778#-NPP0_Yt45mP24xA6WYi:-NPTStRZ5TJuzSjypZKM:bk5hd29) ([raw file](https://github.com/cockroachdb/cockroach/blob/4497047ec68b04c9dd1141a6dbc66684ed8551c2/pkg/sql/catalog/funcdesc/func_desc.go#L865)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think this is a bad message for the validation error -- nothing is being set.
</blockquote></details>

agree very much, changed it to something more non-action.


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented on pkg/sql/catalog/funcdesc/helpers.go:
```diff
@@ -12,74 +12,10 @@ package funcdesc
 
-// VolatilityToProto converts sql statement input volatility to protobuf
-// type.
-func VolatilityToProto(v tree.FunctionVolatility) (catpb.Function_Volatility, error) {
```
These are moved to the new `funcinfo` package.

ajwerner reviewed (approved):
LGTM thanks for the cleanup

chengxiong-ruan commented:
TFTR!
bors r+
